consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,486 @@
+
+E.G:
+// 2f+1 commit messages for the current round and height with a non nil value.
+M ← <Commit, Hc, Rc, V> && |M| = 2f+1 && V != nil
+```
+```
+schedule <function call> after <duration> - This notation schedules the given
+function call to occur after the given duration.
+
+```
+
+### Math notation
+`← - assignment`\
+`= - is equal`\
+`!= - is not equal`\
+`&& - logical and`\
+`|| - logical or`\
+`{X, Y} - the set containing X and Y`\
+`|M| - the cardinality of M`\
+`m ∈ M - m is an element of the set M`\
+`{ m : C(m) } - set builder notiation, the set of messages m such that they satisfy condition C`\
+`∃ m : C(m) - there exists m that satisfies condition C`\
+`∀ m : C(m) - all m satisfy condition C`
+
+### Math Notation examples
+```
+// There exists a commit message m in M such that m's height (Hm) is
+// less than m's round (Rm) and m's value (V) is not important.
+∃ m<Commit, H, R, *> ∈ M : Hm < Rm
+
+// The cardinality of prepare messages in M with height and round equal
+// to Hc and value equal to V is greater than 1 and less than 10.
+1 < |{ m<Prepare, Hc, Rm, Vm> ∈ M : Rm = Hc && Vm = V }| < 10
+```
+
+## Strange things
+
+This is from check message but it doesn't check the message
+it just compares current round to desired
+
+consensus/istanbul/core/backlog.go:64
+  if c.current.Round().Cmp(c.current.DesiredRound()) > 0 {
+
+## Thoughts
+
+The future preprepare timer seems unnecessary, shouldn't the future preprepare
+message simply be handled when moving to the future sequence? Actually I think
+the future prepare timer is there to ensure that the network doesn't race ahead
+of the proposed block times, but it would be better if nodes simply waited some
+amount of time from the last block rather than making a calculation based on
+the time value set by the proposer.
+
+It seems the resetResendRoundChangeTimer functions are there to ensure round
+changes get resent, we need to represent that here because we don't assume
+reliable broadcast.
+
+When a round change timeout occurs it starts the timer for the next round
+change, which will result in another round change message being broadcast for
+the newer round, so why do we need the resendRoundChangeMessage functionality?
+
+I think nodes can get stuck:
+
+Say all nodes are at round 0 they all time out and send round changes for round 1 so they all move to round one.
+
+Now f+1 nodes timeout and send round changes for round 2 they have desired round 2 current round 1. The remaining 2f
+nodes receive the f+1 round changes and set desired round to 2 and schedule
+roundChangeTimeout with a target round of 2.
+
+Now if those 2f nodes have their scheduled roundChangeTimeout execute before
+they have updated their current round to 2 roundChangeTimeout will be a no-op.
+if the round changes sent by these 2f nodes are lost then they will never
+schedule any further round changes without outside intervention.
```

Comment:
I'm not quite following this sequence of events. I agree that you could stall the network if you got into a condition where a node would not schedule a round change timeout & nodes got stuck on different rounds, but I'm not sure that there is a sequence of events that results in roundChangeTimeouts being dropped.

Reply:
So this was more of a thought experiment/working notes. I think in these cases nodes could get stuck in the case that we didn't have the resendRoundChangeTimeout functionality.


Looking at this again I think there is actually a much more straightforward sequence of events that can lead to 
nodes can getting into a situation where they have killed their local chain of round change timeouts.

```
onRoundChangeTimeout(H, R) {
  if H = Hc && R = Rc {
    Rd ← Rc+1
    Sc ← WaitingForNewRound
    schedule onRoundChangeTimeout(Hc, Rd) after roundChangeTimeout(Rd)
    bc<RoundChange, Hc, Rd, PCc>
  }
}
```

So we can see the `onRoundChangeTimeout` function will only schedule another timeout if the internal condition is met. So to kill the local chain of timeouts we just need to not meet that condition once.

So if we imagine that all nodes time out on round 0, the round change timeout condition will be met because when starting a sequence it is scheduled a such `schedule onRoundChangeTimeout(Hc, 0) after roundChangeTimeout(0)`

The next round change will be scheduled with the desired round == 1 but current round is still 0. If all the subsequently broadcast round change messages are lost, then when the scheduled round change timeout runs the expression will not be met  and all nodes will be stuck and never progress from there on.